Troubleshooting.md

```diff
@@ -324,6 +326,8 @@ Note: For newer Gson versions these rules might be applied automatically; make s
 }
 ```
 
+You can also use `<init>(...);` to keep all constructors of that class, but then you might actually rely on JDK Unsafe to create classes without no-args constructor, see [`GsonBuilder.disableJdkUnsafe()`](https://javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/GsonBuilder.html#disableJdkUnsafe()) for more information.
+
 For Android you can add this rule to the `proguard-rules.pro` file, see also the [Android documentation](https://developer.android.com/build/shrink-code#keep-code). In case the class name in the exception message is obfuscated, see the Android documentation about [retracing](https://developer.android.com/build/shrink-code#retracing).
 
-Note: If the class which you are trying to deserialize is actually abstract, then this exception is probably unrelated to R8 and you will have to implement a custom [`InstanceCreator`](https://javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/InstanceCreator.html) or [`TypeAdapter`](https://javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/TypeAdapter.html) which creates an instance of a non-abstract subclass of the class.
+For Android you can alternatively use the [`@Keep` annotation](https://developer.android.com/studio/write/annotations#keep) on the class or constructor you want to keep. That might be easier than having to maintain a custom R8 configuration.

```

Comment:
This is not only for Android. You can also use R8 (or ProGuard) to shrink class files.

Also please change `@Keep` to [`@androidx.annotation.Keep`](https://developer.android.com/reference/androidx/annotation/Keep).

Reply:
Maybe this paragraph is worded a bit misleading then. What I meant was that Android developers can use Android's `@Keep` annotation, which has explicit ProGuard / R8 support by the Android build tooling. Or do ProGuard / R8 in general consider any `@Keep` annotations, regardless of package name and without having to manually configure this first? Do you have any ideas how to better describe this then?

Because this is also written in the context of Android, I am not sure if it is necessary to specify the fully qualified package name. I also preferred linking to the Android Developers page instead of the Javadoc for that `@Keep` annotation because it has links to more information about shrinking.

Reply:
Thanks, linking to DAC is a good idea, and also keeping Android is actually fine, as the annotation is part of an Android library.